File: PerformanceAnalysis/src/main/java/com/android/performanceanalysis/adapter/HomePageAdapter.java
package com.android.performanceanalysis.adapter;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.util.SparseArrayCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.android.performanceanalysis.R;
import com.android.performanceanalysis.data.HomeData;

import java.util.List;


public class HomePageAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final Context mContext;
    private final List<HomeData.ItemView> mItemData;
    private static final int ITEM_TYPE_HEADER = 100000;
    private static final int ITEM_TYPE_TITLE = 111110;
    private static final int ITEM_TYPE_SECOND = 111111;
    private SparseArrayCompat<Integer> mHeaderViews = new SparseArrayCompat<>();

    public HomePageAdapter(Context context, List<HomeData.ItemView> list) {
        this.mContext = context;
        this.mItemData = list;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (mHeaderViews.get(viewType) != null) {
            View v = LayoutInflater.from(mContext).inflate(mHeaderViews.get(viewType), parent,
                    false);
            return new HeadViewHolder(v);
        } else if (ITEM_TYPE_TITLE == viewType) {
            View v = LayoutInflater.from(mContext).inflate(R.layout.activity_home_page_title,
                    parent, false);
            return new TitleViewHolder(v);
        } else {
            View v = LayoutInflater.from(mContext).inflate(R.layout.activity_home_page_item,
                    parent, false);
            return new ItemViewHolder(v);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        int viewType = getItemViewType(position);
        if (mHeaderViews.get(viewType) != null) {
            HeadViewHolder headViewHolder = (HeadViewHolder) viewHolder;
            headViewHolder.scan.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {

                }
            });
            return;
        } else {
            position -= getHeadersCount();
        }
        if (viewType == ITEM_TYPE_TITLE) {
            TitleViewHolder titleViewHolder = (TitleViewHolder) viewHolder;
            titleViewHolder.title.setText(mItemData.get(position).desc);
        } else {
            ItemViewHolder itemViewHolder = (ItemViewHolder) viewHolder;
            itemViewHolder.name.setText(mItemData.get(position).desc);
//            itemViewHolder.icon.setText(mItemData.get(position).icon);
            final int pos = position;
            itemViewHolder.name.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    switch (pos) {
                        case 1:

                            break;
                        case 2:

                            break;
                        case 3:

                            break;
                        default:
                            break;
                    }

                }
            });
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (isHeaderViewPos(position)) {
            return mHeaderViews.keyAt(position);
        } else if (isTitle(position)) {
            return ITEM_TYPE_TITLE;
        } else {
            return ITEM_TYPE_SECOND;
        }
    }

    @Override
    public int getItemCount() {
        return mItemData.size() + getHeadersCount();
    }

    private boolean isHeaderViewPos(int position) {
        return position < getHeadersCount();
    }

    private int getHeadersCount() {
        return mHeaderViews.size();
    }

    public void addHeaderView(int view) {
        mHeaderViews.put(mHeaderViews.size() + ITEM_TYPE_HEADER, view);
    }

    public void removeHeaderView() {
        mHeaderViews.clear();
    }

    public boolean isHaveHeaderView() {
        return mHeaderViews.size() > 0;
    }

    private boolean isTitle(int position) {
        return "".equals(mItemData.get(position - getHeadersCount()).icon);
    }

    class ItemViewHolder extends RecyclerView.ViewHolder {
        TextView name;
        TextView icon;

        @SuppressLint("WrongViewCast")
        public ItemViewHolder(View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.tv_item_name);
            icon = itemView.findViewById(R.id.tv_item_icon);
        }
    }

    class TitleViewHolder extends RecyclerView.ViewHolder {
        TextView title;

        public TitleViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.tv_title);
        }
    }

    class HeadViewHolder extends RecyclerView.ViewHolder {
        TextView scan;

        public HeadViewHolder(View itemView) {
            super(itemView);
            scan = itemView.findViewById(R.id.tv_scan);
        }
    }
}

File: Architecture/src/main/java/com/android/architecture/adapter/HomePageAdapter.kt
package com.android.architecture.adapter

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.support.v4.util.SparseArrayCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.android.architecture.R
import com.android.architecture.data.HomeData
import com.android.architecture.demolist.lifecycle.LifecycleActivity
import com.android.architecture.demolist.livedata.LiveDataActivity
import com.android.architecture.demolist.navigation.NavigationActivity
import com.android.architecture.demolist.viewmodel.ViewModelActivity

class HomePageAdapter(private val mContext: Context, private val mItemData: List<HomeData.ItemView>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val mHeaderViews = SparseArrayCompat<Int>()

    private val headersCount: Int
        get() = mHeaderViews.size()

    val isHaveHeaderView: Boolean
        get() = mHeaderViews.size() > 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        if (mHeaderViews.get(viewType) != null) {
            val v = LayoutInflater.from(mContext).inflate(mHeaderViews.get(viewType)!!, parent,
                    false)
            return HeadViewHolder(v)
        } else if (ITEM_TYPE_TITLE == viewType) {
            val v = LayoutInflater.from(mContext).inflate(R.layout.activity_home_page_title,
                    parent, false)
            return TitleViewHolder(v)
        } else {
            val v = LayoutInflater.from(mContext).inflate(R.layout.activity_home_page_item,
                    parent, false)
            return ItemViewHolder(v)
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        var position = position
        val viewType = getItemViewType(position)
        if (mHeaderViews.get(viewType) != null) {
            val headViewHolder = viewHolder as HeadViewHolder
            headViewHolder.scan.setOnClickListener { }
            return
        } else {
            position -= headersCount
        }
        if (viewType == ITEM_TYPE_TITLE) {
            val titleViewHolder = viewHolder as TitleViewHolder
            titleViewHolder.title.text = mItemData[position].desc
        } else {
            val itemViewHolder = viewHolder as ItemViewHolder
            itemViewHolder.name.text = mItemData[position].desc
            //            itemViewHolder.icon.setText(mItemData.get(position).icon);
            val pos = position
            itemViewHolder.name.setOnClickListener {
                val intent = Intent()
                when (pos) {
                    1 -> {
                        intent.setClass(mContext, LifecycleActivity::class.java)
                        mContext.startActivity(intent)
                    }
                    2 -> {
                        intent.setClass(mContext, ViewModelActivity::class.java)
                        mContext.startActivity(intent)
                    }
                    3 -> {
                        intent.setClass(mContext, LiveDataActivity::class.java)
                        mContext.startActivity(intent)
                    }
                    4 -> {

                    }
                    5 -> {

                    }
                    6 -> {
                        intent.setClass(mContext, NavigationActivity::class.java)
                        mContext.startActivity(intent)
                    }
                    7 -> {

                    }
                    8 -> {

                    }
                    9 -> {

                    }
                    10 -> {

                    }
                    else -> {
                    }
                }
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (isHeaderViewPos(position)) {
            mHeaderViews.keyAt(position)
        } else if (isTitle(position)) {
            ITEM_TYPE_TITLE
        } else {
            ITEM_TYPE_SECOND
        }
    }

    override fun getItemCount(): Int {
        return mItemData.size + headersCount
    }

    private fun isHeaderViewPos(position: Int): Boolean {
        return position < headersCount
    }

    fun addHeaderView(view: Int) {
        mHeaderViews.put(mHeaderViews.size() + ITEM_TYPE_HEADER, view)
    }

    fun removeHeaderView() {
        mHeaderViews.clear()
    }

    private fun isTitle(position: Int): Boolean {
        return "" == mItemData[position - headersCount].icon
    }

    internal inner class ItemViewHolder @SuppressLint("WrongViewCast")
    constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var name: TextView
        var icon: TextView

        init {
            name = itemView.findViewById(R.id.tv_item_name)
            icon = itemView.findViewById(R.id.tv_item_icon)
        }
    }

    internal inner class TitleViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var title: TextView

        init {
            title = itemView.findViewById(R.id.tv_title)
        }
    }

    internal inner class HeadViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var scan: TextView

        init {
            scan = itemView.findViewById(R.id.tv_scan)
        }
    }

    companion object {
        private val ITEM_TYPE_HEADER = 100000
        private val ITEM_TYPE_TITLE = 111110
        private val ITEM_TYPE_SECOND = 111111
    }
}
